Add netCDF PIO capability for restarts and run-time history for dev/ufs-weather-model

DeniseWorthen opened the pull request:
# Pull Request Summary

Enable netCDF restarts and run-time netCDF history using PIO.

## Description
<!--
Provide a detailed description of what this PR does.
What bug does it fix, or what feature does it add?
Is a change of answers expected from this PR?

Please also include the following information: 
* Add any suggestions for a reviewer 
* Mention any labels that should be added:  _bug_, _documentation_, _enhancement_, _new feature_
* Are answer changes expected from this PR? Please describe the changes and the reason why in addition to which of the following labels would apply: _mod_def change_, _out_grd change_, _out_pnt change_, _restart file change_, _Regression test_ 
-->

1. PIO configuration

PIO is configured via configuration settings in WAV attributes for UFS and via share code for CESM. Defaults are provided for the stride, number of io tasks and rearranger settings. The PIO subsystem is managed and initialized in ``wav_pio_mod.F90``

2. netCDF restarts

PIO/netCDF restarts are enabled by using ``use_restartnc=true`` in the configuration attributes, otherwise native WW3 binary restarts will be written. If netCDF restarts are enabled, an additional option exists to restart from an existing binary restart (``restart_from_binary=true``). In this case, the binary restart filename needs the same form as the netCDF restarts (minus the .nc at the end). 

With netCDF restarts enabled, restarts will be written at the ``rstwr`` frequency defined in the cap, which will allow both the flexible (ie, non-interval) restarts and end-of-run restarts to be enabled in the same way as for other coupled components in UFS. These two features are not implemented in this work. 

The netCDF restarts are read and written in a new routine ``wav_restart_mod.F90``. Only two fields (``va`` and ``mapsta``) are written to the restart file by default. As noted in the comment in #1298, when waves are in the slow-coupling loop, the ice field is also written. The addition of extra restart fields is generalized via the already-existing ``restart%extra`` namelist. 

3. netCDF history

Runtime netCDF history is enabled by using ``use_historync=true`` in the configuration attributes, otherwise native WW3 binary history files will be written. The PIO/netCDF capability builds on the existing serial netCDF capability in the mesh cap. The two existing routines used for this (``w3iogoncdmd.F90, wav_grdout.F90``) have been reduced to a new single netCDF history module ``wav_history_mod.F90`` containing the same functionality. 

History files can be custom named (if ``user_histname = true``), otherwise they will be named in the native WW3 file naming. History files will be written at the ``histwr`` frequency. This frequency will be set using the normal ``history_n,history_option`` settings, if present. Otherwise, the ``field%stride`` frequency will be used to determine the ``histwr`` frequency. (This requires that the stride be set in units of seconds, which is less flexible than enabling the history_n,history_option settings.)

4. Removal of mesh-cap specific modifications to ``w3wavemd, w3iorsmd and w3iogomd.``. 
 
Modifications to the three listed routines were required for the mesh cap to allow for additional flexibility of restart and history writing. The changes in w3iogomd were resolved w/ the PR to dev/ufs-weather-model for the Langmuir turbulence calculations (#1034). With the new restartnc and historync options via PIO, the changes to the other two routines (wavemd and iorsmd) are no longer required, meaning these two routines are now much more aligned with the same code in the develop branch. Since the dev/ufs-weather-model branch has not been updated for nearly a year, the modifications are best seen by comparing these two routines in this PR branch against the develop branch at 4d8c3156


### Performance Testing 

To test the performance of the new PIO/netCDF restarts vs the existing binary restarts, a series of runs were made using a DATM + 1/4 deg MOM6+CICE6 and WW3 utilizing the ``uglo_15km-hr4_BlkSlake_360`` mesh (~1.8M spatial elements) with 1800 frequencies + directions. Although this PR also enables PIO to write the run-time netCDF history files (vs the current serial netCDF implementation), no performance testing was done for the history writes.

Cases were run for 24 hours, with all component output turned off except for WW3 restarts, which were written every 2 hours. A range of resources was tested for WW3, from ~400 to ~2800. The primary metric used for evaluation was the mean time for a Model Advance on non-restart intervals vs restart intervals (the restart-write "penalty"). Because WW3 is cold-starting, the first 6 hours of integration was ignored. 

The primary configuration options for PIO are the stride (and/or number of IO tasks) and the rearranger.  In addtion to these options, two options were tested for writing the ``VA`` field---either as a single array, dimensioned (1.8M,1800) or with each spectral index written to a separate, named field (e.g. ``va.0001`` through ``va.1800``). This "multi-field" option creates an identically sized restart file and has been also tested in UWM, where it passes all tests. Thanks go to @jedwards4b for helping me get the single array working for subset rearranger and the operational mesh, which allowed me to test the single field vs multi-field options.

Initial scout runs were made to identify the best combination of stride and rearranger and va-array structure. At all WAV tasks tests, the best results were consistently found using ``pio_stride=4``, ``pio_rearranger=subset`` and with the ``VA`` field written as multiple fields. The finding that stride=4 was a little unexpected, since by default, as resources increase the stride is also set to increase. I think that because we are under-subscribing the nodes in order to provide enough memory to WW3, this acts as a sort of de-facto striding and stride=4 (and even stride=2) can be used even at high task counts. 

Once the best combination of parameters were found for PIO, cases were run for a total of 3 times. The binary case was also run three times and all runs were averaged. The raw results can be found [here](https://docs.google.com/spreadsheets/d/1qAH_1ge8SYTnMTvxmGRgyllUernoWrDHq4xDvUoqRcM/edit?usp=sharing)  

The summary results are presented below. 

<img width="600" alt="Screenshot 2024-10-14 at 10 21 03 AM" src="https://github.com/user-attachments/assets/d13f5fe1-f298-4992-930f-df56c94937f8">
<img width="600" alt="Screenshot 2024-10-14 at 8 17 07 AM" src="https://github.com/user-attachments/assets/06aef07c-e56f-4dc1-87c3-3fdcedf6823b">
<img width="600" alt="Screenshot 2024-10-14 at 8 17 18 AM" src="https://github.com/user-attachments/assets/6f69c655-3cb7-4724-beff-d4d6269eb29f">
<img width="600" alt="Screenshot 2024-10-14 at 8 17 25 AM" src="https://github.com/user-attachments/assets/aa2b0dee-51a6-4d52-bb68-a72118374e40">


### Issue(s) addressed
<!--
* Please list any issues associated with this PR, including those the PR will fix/close. For example:  
- fixes #<issue_number>
- fixes noaa-emc/ww3/issues/<issue_number>
-->
- fixes #1298 

### Commit Message
<!--
Please provide a short summary of this PR, which will be used during _Squash and Merge_ and will be shown as a git log message.  Be sure to add any co-authors here. 
-->

### Check list  

<!-- After creating the PR you can check each of the items below that have been completed -->

- [ ] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [ ] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [ ] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [ ] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested?

Using UFS-WM, a baseline was created and verified to pass for all tests which include a WAV component. Full testing as well as scalability results will be documented in the associated UFS-WM PR.

DeniseWorthen committed:
clean branch for pio work

DeniseWorthen committed:
add new files

DeniseWorthen committed:
rename w3iogoncmd as wav_history_mod

DeniseWorthen committed:
rework order to enable PIO initialization

* remove wav_grdout routine, now moved into wav_history_mod

DeniseWorthen committed:
clean up w3iors, ascii and config variables

* remove more cap stuff from w3iorsmd. only ww3 native filenaming
is possible w/ binary restarts
* remove ifdef w3_ascii from w3wavemd, since the ascii commit is not
yet present in mesh cap branch
* clean up config variable logic for filenaming

DeniseWorthen committed:
prevent mpi_startalls if restartnc is active

* nrqrs can be non-zero if also using the restart from binary
option

DeniseWorthen committed:
move setting of flpart flag

* flpart needs to be set for either historync or not

DeniseWorthen committed:
remove unneeded CESMCOUPLED from w3iorsmd

DeniseWorthen committed:
update nc block location, add log messages

* move restart and history nc blocks outside of ww3 time testing
block.
* add log print messages for reading and writing restarts

DeniseWorthen committed:
use pio_ioformat for create mode

DeniseWorthen committed:
make iosys a pointer, add debug level config

DeniseWorthen committed:
fix ww3 thrd cases

DeniseWorthen committed:
remove stray bin2nc ifdef

DeniseWorthen committed:
fixes for logs and wav_history

* get logging working correctly for ufs
* fix noclobber variable and file name in wav_history
* clarify some comments

DeniseWorthen committed:
resolve issue for binary history files

* make binary history files match when restartnc=true
* rework wav_restart_mod, which originally was designed to be able
to read and write restarts for testing purposes from inside wav_comp_nuopc.

DeniseWorthen committed:
fix multiple writes when waves are in slow loop

DeniseWorthen committed:
rename oldst2 variable in wav_restart

DeniseWorthen committed:
add verbose log option

* verboselog is true by default, but can be set false by config
* the header text for ww3 logging as it steps through time is now
turned off in w3init and placed into the mesh cap. this allows the
mesh cap to order the logging correctly

DeniseWorthen committed:
add optional restart files via nml

DeniseWorthen committed:
resolve issue w/ nml for extra fields

* move block where addrstflds was set to before call to w3init
since restarts are read in w3init
* ensure that if nml file lacks a specification of extra fields,
the default value of "unset" will not be returned as a field name

DeniseWorthen committed:
add generic read/write for added restart fields

* only ice is added for now
* tab cleanup in w3grid

DeniseWorthen committed:
fix generic read/write in wav_restart

* need to send explicit array bounds for ice array since
it is 0:nsea
* all baselines b4b against f9531d0

DeniseWorthen committed:
initialize variables

* intialize floutg and floutg2 which are can be unintialized when
waves are in slow loop and historync is true

DeniseWorthen commented:
@MatthewMasarik-NOAA Can I ask if you've tested the mesh you provided me in debug mode? 

When using the dev/ufs-weather-model branch, I am getting a failure in debug mode at LN 1388 in w3init

https://github.com/NOAA-EMC/WW3/blob/7f548c795a348bbb0fe4967dd25692c79036dc73/model/src/w3initmd.F90#L1384-L1398

MatthewMasarik-NOAA commented:
Hi @DeniseWorthen, I have not tested this mesh in either standard or debug mode.  I was provided this HR4 mesh from @AliS-Noaa and @JessicaMeixner-NOAA who have done the unstructured grid creation, and could probably give that information.  

You had mentioned offline that at one point that Ali had told you a grid was passing in standalone, but not in coupled.  I had not been briefed on that. Do you know if anyone followed up on that at the time?  Or can you confirm if this new mesh is the same as the one you were discussing with Ali?

DeniseWorthen commented:
@MatthewMasarik-NOAA The remark about failing in coupled mode was from an informal lunch-time discussion I had w/ Ali, some time ago, where he simply mentioned in passing what he was working on. It was hardly a "briefing" I had on the subject. 

I have only the mesh you provided me, as well as one which Jiande gave me which he is currently using for HR4.

MatthewMasarik-NOAA commented:
> @MatthewMasarik-NOAA The remark about failing in coupled mode was from an informal lunch-time discussion I had w/ Ali, some time ago, where he simply mentioned in passing what he was working on. It was hardly a "briefing" I had on the subject.
> 
> I have only the mesh you provided me, as well as one which Jiande gave me which he is currently using for HR4.

@DeniseWorthen, understood.  It was more than has been shared with me so you know more than I do at this point.

JessicaMeixner-NOAA commented:
I think we might be getting meshes confused as well.  HR4 was supposed to originally have an update and use the new mesh from @AliS-Noaa - however, due to time it's using the same as HR3.   A new HR with updated wave grids will happen.  @sbanihash I think @DeniseWorthen has pointed out an oversight from us that we should test this new mesh with debug mode on - would you be able to assign someone to run the new mesh in debug mode?

DeniseWorthen commented:
I believe I had the same failure w/ the HR4 files that Jiande provided me. But I will confirm and post back either yes or no.

DeniseWorthen commented:
My run directory on Gaea is

/gpfs/f5/nggps_emc/scratch/Denise.Worthen/ww3pio/datm.gfsv17

grep F90 err | grep w3
```
2:200:  95: fv3_s2sw.datm.dev  0000000001FCA864  w3initmd_mp_w3ini        1388  w3initmd.F90
11:370:  81: fv3_s2sw.datm.dev  0000000001FC9AE3  w3initmd_mp_w3ini        1388  w3initmd.F90
20:542:  93: fv3_s2sw.datm.dev  0000000001FC9E5E  w3initmd_mp_w3ini        1388  w3initmd.F90
....
```

In this case, I'm using a mod_def created with the 'gefs' settings in the grid.inp file. 

This is a test case I built w/ a DATM+MOM+CICE. See /gpfs/f5/nggps_emc/scratch/Denise.Worthen/ww3pio/datm.500k for a test case that runs w/ the same executable but w/ a different mesh and mod_def (the global 500k mesh I had used during mesh cap development).

DeniseWorthen commented:
A couple of updates. One, I've been able to create a mod_def file in Release mode by setting (on hercules) in my shell

```
ulimit -s unlimited
``` 

Two, I've been able to run both the mesh Matt provided as well as the HR4 mesh in debug mode on Gaea. The trick was to reduce the processor count but over-subscribe the nodes, ie, running the WAV model on 512 PE but spreading the entire job over 1024 processors. Just setting high processor counts appears to cause the model to hang or seg-fault. 

I suspect memory use is not that optimal---for example in triadmd, cellvertex is statically allocated but is used only locally. It is dimensioned as (1.8M,8,2).

MatthewMasarik-NOAA commented:
> A couple of updates. One, I've been able to create a mod_def file in Release mode by setting (on hercules) in my shell
> 
> ```
> ulimit -s unlimited
> ```
> 
> Two, I've been able to run both the mesh Matt provided as well as the HR4 mesh in debug mode on Gaea. The trick was to reduce the processor count but over-subscribe the nodes, ie, running the WAV model on 512 PE but spreading the entire job over 1024 processors. Just setting high processor counts appears to cause the model to hang or seg-fault.
> 
> I suspect memory use is not that optimal---for example in triadmd, cellvertex is statically allocated but is used only locally. It is dimensioned as (1.8M,8,2).

I'm very happy to hear that is the case Denise.  Setting the ulimit and prescribing 2-cores per tasks just for memory are what I needed to do in the standalone case for heavier meshes.

DeniseWorthen commented:
> I'm very happy to hear that is the case Denise. Setting the ulimit and prescribing 2-cores per tasks just for memory are what I needed to do in the standalone case for heavier meshes.

Perhaps sharing that info would have saved me some time then.

MatthewMasarik-NOAA commented:
> Perhaps sharing that info would have saved me some time then.

Again, Denise, I did not do the testing for this mesh.  I understood it to be similar weight to the previous (15km) mesh, and had no experience with this particular mesh.  Always happy to work collaboratively, though.

JessicaMeixner-NOAA commented:
> > Perhaps sharing that info would have saved me some time then.
> 
> Again, Denise, I did not do the testing for this mesh. I understood it to be similar weight to the previous (15km) mesh, and had no experience with this particular mesh. Always happy to work collaboratively, though.

@MatthewMasarik-NOAA you did scaling runs in standalone mode for both of these meshes.    @DeniseWorthen I'm so sorry this information was not effectively communicated with you.  I am hoping to find time tomorrow to find my set-ups that I ran in coupled mode to share.  Please let me know if there is any other information I can look for during this time to help.

DeniseWorthen committed:
replace statically allocated arrays to reduce memory

* local arrays dimension nx or greater are locally allocated
* deallocations added, even though not required for user clarity

DeniseWorthen commented:
I've been able to write a netCDF restart for the ``uglo_15km-hr4_BlkSlake_360.ww3.msh`` mesh. It is 13GB. I'm running on 512tasks (16 nodes) in my DATM-S2SW setup on gaea.

In my feature/pio4ww3, I'm currently testing the reduction of statically allocated arrays in triamd.F90. You can see the corresponding changes against the develop branch [here](https://github.com/NOAA-EMC/WW3/compare/develop...DeniseWorthen:WW3:feature/devmemuse_triamd?expand=1)

MatthewMasarik-NOAA commented:
> I've been able to write a netCDF restart for the `uglo_15km-hr4_BlkSlake_360.ww3.msh` mesh. It is 13GB. I'm running on 512tasks (16 nodes) in my DATM-S2SW setup on gaea.
> 
> In my feature/pio4ww3, I'm currently testing the reduction of statically allocated arrays in triamd.F90. You can see the corresponding changes against the develop branch [here](https://github.com/NOAA-EMC/WW3/compare/develop...DeniseWorthen:WW3:feature/devmemuse_triamd?expand=1)

That's great news @DeniseWorthen.  Thank you for the update.  I'm looking through the w3triadmd changes you made. Also, I can confirm that the restarts I'm creating for HR4 on Hera are also 13GB.

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/reducememuse_triamd' into feature/pio4ww3

DeniseWorthen committed:
in wav_restart, make arrays allocatable

* global_input and _output can be allocatable

DeniseWorthen committed:
write va in nspec loop

DeniseWorthen committed:
fix compile error

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/pio4ww3

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/debugsubset

DeniseWorthen committed:
make dofs i*8 arrays

DeniseWorthen committed:
whitespace cleanup

DeniseWorthen committed:
write/read each nspec separately

DeniseWorthen committed:
fix dimension error

DeniseWorthen committed:
fix calculation of dof3d for large meshes

DeniseWorthen commented:
@JacobCarley-NOAA @junwang-noaa @GeorgeVandenberghe-NOAA @CatherineThomas-NOAA I've updated the PR description above with the results of my performance testing. I'm tagging you all here in case you don't watch this repo.

junwang-noaa commented:
@DeniseWorthen Thanks for the testing. Would you please also check the max time instead of mean time for ww3?

DeniseWorthen commented:
@junwang-noaa If I look at the time-series, the maximum time will most likely always be at hour=6, since WW3 gets a bit faster as it spins up. I'm not sure max gives a good measure of the cost of a restart write in this case.

DeniseWorthen committed:
Merge branch 'feature/debugsubset_multifld' into feature/pio4ww3

JacobCarley-NOAA commented:
@DeniseWorthen excellent work - thanks for tagging me. The new capability seems to show a nice improvement. Do you have performance numbers for the overall runtime? (e.g. for whatever a typical forecast length is in operations).

DeniseWorthen commented:
@JacobCarley-NOAA For these tests, I was using a DATM configuration in order to focus solely on WW3 restart performance. I have not done any testing w/ the actual operational configuration. I think George V. has been doing that testing.

junwang-noaa commented:
@DeniseWorthen Let me clarify if I understand correctly. By "max time" I mean the "Max" time in ESMF_Profile.summary, usually that time gives the total run time of a model component.

GeorgeVandenberghe-NOAA commented:
I have.  With the hr4 testcase I got from Ruiyu, enabling restarts adds 800
seconds to a 72 hour forecast with 7 restarts (12 hour restart  interval)
on WCOSS2.    I will note again the Ruyiu testcase does not run on hercules
or orion.. hangs after wave model is reported as starting.   This is NOT
evidence the issue is in wave.   There is a second point output issue with
WAVE which adds 2000 seconds to startup once but that is known and being
addressed so I am not worried about it and have reduced point outputs to
almost none to get rid of it for now.


On Tue, Oct 15, 2024 at 8:57 AM Denise Worthen ***@***.***>
wrote:

> @JacobCarley-NOAA <https://github.com/JacobCarley-NOAA> For these tests,
> I was using a DATM configuration in order to focus solely on WW3 restart
> performance. I have not done any testing w/ the actual operational
> configuration. I think George V. has been doing that testing.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/WW3/pull/1303#issuecomment-2413844684>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FSBYDY25PAJRKHYKVTZ3UGMPAVCNFSM6AAAAABOYROF42VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMJTHA2DINRYGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

George W Vandenberghe

*Lynker Technologies at * NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***


301-683-3769(work)  3017751547(cell)

DeniseWorthen commented:
> @DeniseWorthen Let me clarify if I understand correctly. By "max time" I mean the "Max" time in ESMF_Profile.summary, usually that time gives the total run time of a model component.

When Gaea-C5 comes back from maintenance, I will add that information.

CatherineThomas-NOAA commented:
Thanks for the tag @DeniseWorthen.  Tagging @RuiyuSun as well.

sbanihash commented:
@DeniseWorthen awesome work! This is very exciting. Thanks for adding the run details to the PR description.

junwang-noaa commented:
@DeniseWorthen Is your branch https://github.com/DeniseWorthen/WW3/tree/feature/pio4ww3 ready for testing in HR4? If yes, @GeorgeVandenberghe-NOAA can you use Denise's branch to test ww3 IO performance in HR4? Thanks

DeniseWorthen commented:
@junwang-noaa I am putting in a final change, which is to error out if the dimensions in the netcdf restart file don't match. It should be finished by today. 

There will need to be changes in ``ufs.configure`` in order for George to test. To tell you the right settings, I need to know whether HR4 will have a binary WW3 restart available to warm start or is it a cold start?

DeniseWorthen committed:
add error checks for dimensions in restart file

* tested by using nco to alter the nth,nk values in a restart
file. The model correctly aborted after finding nth did not
match the expected value

jkbk2004 commented:
@MatthewMasarik-NOAA we are about to merge https://github.com/ufs-community/ufs-weather-model/pull/2463. We consider to work on this PR after the PR 2463. Let me know if you need more time to review this PR.

MatthewMasarik-NOAA commented:
> @MatthewMasarik-NOAA we are about to merge [ufs-community/ufs-weather-model#2463](https://github.com/ufs-community/ufs-weather-model/pull/2463). We consider to work on this PR after the PR 2463. Let me know if you need more time to review this PR.

Hi @jkbk2004. I have the tests going on hera right now, and am working on the code review.  It is a significant PR though, so I could use a little more time to review if that's OK for your schedule.

DeniseWorthen committed:
fix deallocation issue when dbug_flag>5

* delay the deallocation of gindex arrays until after the diagnose_mesh
routine is called. This only impacts cases where dbug_flag>5, compiled
with gnu+debug

MatthewMasarik-NOAA commented:
The RTs just came back and here is the [test_changes.list](https://github.com/user-attachments/files/17482717/test_changes.list.txt).  The tests were run before the latest commit, and from a quick look at the list, the changes might all be explained by that.  I'll have a closer look tomorrow.

DeniseWorthen commented:
The test_changes.list is not related to the latest commit---that fixes only an issue if dbug_flag>5 and dbug_flag=0 for all tests. 

The test_changes.list is because we do not compare ww3 restart files for any cpld test, because ww3 restart files do not reproduce. All of tests should report the reason for failure is that the restart file (prefix.ww3.r.date.nc) does not exist in the baseline. Let me know if that is not the case.

MatthewMasarik-NOAA commented:
> The test_changes.list is not related to the latest commit---that fixes only an issue if dbug_flag>5 and dbug_flag=0 for all tests.
> 
> The test_changes.list is because we do not compare ww3 restart files for any cpld test, because ww3 restart files do not reproduce. All of tests should report the reason for failure is that the restart file (prefix.ww3.r.date.nc) does not exist in the baseline. Let me know if that is not the case.

Okay, thanks for that information.  I suppose I should re-run the tests at any rate after the new commit.  Do you anticipate any more commits?

DeniseWorthen commented:
I will run against the UWM baseline on hercules overnight and post the test_changes.list.

MatthewMasarik-NOAA commented:
I ran the RTs last night on hera [RegressionTests_hera.log](https://github.com/user-attachments/files/17497666/RegressionTests_hera.log.txt), [test_changes.list](https://github.com/user-attachments/files/17497667/test_changes.list.txt). Most on the change list are failures due to comparision of wave restarts (which do not pass), as @DeniseWorthen pointed out. A few are 'restart' tests that failed to start, so that is expected as well.  There are only two changes then remaining: `cpld_control_gfsv17_iau` and `cpld_mpi_pdlib_p8_intel`.   The `cpld_control_gfsv17_iau` is likely due to the reported issue.  So `cpld_mpi_pdlib_p8_intel` is the only changed test that I could not attribute a reason for the failure, and the reason given was `UNABLE TO START TEST`.  @DeniseWorthen I wanted to check with you to see if you are seeing the same thing, and if there is a reason for the `cpld_mpi_pdlib_p8_intel` test to fail that I overlooked or wasn't aware of?

DeniseWorthen commented:
I think the issue there is that currently in the rt.conf, ``cpld_mpi_pdlib_p8`` (mistakenly) is dependent on ``cpld_control_pdlib_p8``. So if that test does not complete/pass, the mpi test won't run. 

The iau test did not run for the same reason. The test it is dependent on (``cpld_control_gfsv17``) did not pass, so it did not run.

MatthewMasarik-NOAA commented:
> I think the issue there is that currently in the rt.conf, `cpld_mpi_pdlib_p8` (mistakenly) is dependent on `cpld_control_pdlib_p8`. So if that test does not complete/pass, the mpi test won't run.
> 
> The iau test did not run for the same reason. The test it is dependent on (`cpld_control_gfsv17`) did not pass, so it did not run.

Great, thank you for giving that clarification.  I'm okay with the testing then.  I'll work on finishing the code review today.  If everything goes well it should be ready late today or early tomorrow.  Fyi @jkbk2004

jkbk2004 commented:
@MatthewMasarik-NOAA all sounds good. we can go ahead to commit for full testing this afternoon. @DeniseWorthen can you attach hera pre-test log and push test_changes.list at https://github.com/ufs-community/ufs-weather-model/pull/2445 ?

DeniseWorthen commented:
@jkbk2004 The test_changes list was generated on Hercules and posted at [b5ae65b](https://github.com/ufs-community/ufs-weather-model/pull/2445/commits/b5ae65b035142bee8c3d82b954fdabdda58f7992). However, the WW3 PR has still not been approved, so we should not schedule it for commit until it is.

MatthewMasarik-NOAA reviewed (approved):
**Code review**
Pass

**Testing**
Pass

See logs attached to https://github.com/NOAA-EMC/WW3/pull/1303#issuecomment-2433359469 and description in https://github.com/NOAA-EMC/WW3/pull/1303#issuecomment-2433406834 as well.

Approved.

MatthewMasarik-NOAA commented:
Fyi @jkbk2004 @DeniseWorthen 

thanks for your patience

FernandoAndrade-NOAA commented:
Testing on #[2445](https://github.com/ufs-community/ufs-weather-model/pull/2445) has completed, please continue with the merge process.

MatthewMasarik-NOAA commented:
> Testing on #[2445](https://github.com/ufs-community/ufs-weather-model/pull/2445) has completed, please continue with the merge process.

Will do.

MatthewMasarik-NOAA commented:
@DeniseWorthen thank you for your hard work developing and testing this very fine upgrade to WW3.
